go/internal/services/usagelimiter/limit.go

```diff
@@ -0,0 +1,46 @@
+package usagelimiter
+
+import (
+	"context"
+	"database/sql"
+	"math"
+
+	"github.com/unkeyed/unkey/go/pkg/db"
+	"github.com/unkeyed/unkey/go/pkg/otel/tracing"
+)
+
+func (s *service) Limit(ctx context.Context, req UsageRequest) (UsageResponse, error) {
+	ctx, span := tracing.Start(ctx, "usagelimiter.Limit")
+	defer span.End()
+
+	limit, err := db.Query.FindKeyCredits(ctx, s.db.RW(), req.KeyId)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return UsageResponse{Valid: false, Remaining: 0}, nil
+		}
+
+		return UsageResponse{Valid: false, Remaining: 0}, err
+	}
+
+	if !limit.Valid {
+		return UsageResponse{Valid: true, Remaining: -1}, nil
+	}
+
+	remaining := limit.Int32
+	// Key doesn't have enough credits to cover the request cost
+	if remaining <= 0 && req.Cost != 0 || remaining-req.Cost < 0 {
+		return UsageResponse{Valid: false, Remaining: 0}, nil
+	}
+
+	remaining = int32(math.Max(float64(0), float64(remaining-req.Cost)))
+	err = db.Query.UpdateKeyCredits(ctx, s.db.RW(), db.UpdateKeyCreditsParams{
+		ID:        req.KeyId,
+		Operation: "set",
```

Comment:
this smells like a race condition. we should subtract at the db layer I think.

Consider 2 limit request coming in for the same key
1. A reads current credits at 10
2. B reads current credits at 10
3. A sets credits to 9 
4. B sets credits to 9

We have last write wins scenario here, instead of correctly updating the credits to 8 after everything went through

Reply:
yeah agreed. the update credits does this already